### PR TITLE
[Follow-up] Polish keyboard navigation to hyperlinks

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request
Polishes #13405 with some additional feedback found in testing and post-mortem reviews. Such feedback includes:
- [x] ControlInteractivity vs ControlCore split ([link](https://github.com/microsoft/terminal/pull/13405#discussion_r919365435))
- [x] clearing the selection should be under lock when copying text via "Enter"
- [x] move mark mode keybindings into a helper function
- [x] decide if "Enter" should be configurable or non-configurable ([link](https://github.com/microsoft/terminal/pull/13405#discussion_r919379305))
- [x] rename `_isTargetingUrl`
- [x] bugfix: ctrl+enter when the link is outside of the viewport

## References
Original PR: #13405
Relevant issue: #6649
Epic: #4993

### CODE REVIEW

**Comment by zadjii-msft** [comment]
Marking for discussion, but drill into the links above folks. The discussion is happening on the original PR.

**Comment by zadjii-msft** [comment]
> * [x]  ControlInteractivity vs ControlCore split ([link](https://github.com/microsoft/terminal/pull/13405#discussion_r919365435))

Meh it's fine

> * [x]  decide if "Enter" should be configurable or non-configurable ([link](https://github.com/microsoft/terminal/pull/13405#discussion_r919379305))

Leave it as always a feature of Mark Mode, and add it back as a possible setting, so enter on a mouse selection will work, but you can `unbound` it and it'll stop working for a mouse selection, but still work in mark mode.

**Comment by carlos-zamora** [comment]
FYI: this will have a merge conflict with #13659

**Review by lhecker — APPROVED** [review]

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 1]
```diff
@@ -587,24 +587,53 @@ std::wstring Terminal::GetHyperlinkAtViewportPosition(const til::point viewportP
...
+        //       so we need to cleverly set the y-pos to 0.
+        const til::point viewportPos{ bufferPos.x, 0 };
+        const auto results = patterns.findOverlapping(viewportPos, viewportPos);
+        if (!results.empty())
+        {
+            result = results.front();
+            result->start.y += bufferPos.y;
+            result->stop.y += bufferPos.y;
+        }
+    }
```
Comment (lhecker) [inline src/cascadia/TerminalCore/Terminal.cpp:629]: Why do we have to handle hyperlinks outside of the viewport?

Comment (DHowett) [reply src/cascadia/TerminalCore/Terminal.cpp:629]: Keyboard-based navigation through the buffer looking for links would not be very effective if it was constrained to only working in the visible region :)

Comment (lhecker) [reply src/cascadia/TerminalCore/Terminal.cpp:629]: Oh I guess the scrolling is happening in the caller then?

**Review by zadjii-msft — APPROVED** [review]

**Comment by carlos-zamora** [comment]
@msftbot merge this in 10 minutes

**Review by DHowett — CHANGES_REQUESTED** [review]
Blocking til after my meeting!

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 2]
```diff
@@ -587,24 +587,53 @@ std::wstring Terminal::GetHyperlinkAtViewportPosition(const til::point viewportP
...
+        {
+            result->start = _ConvertToBufferCell(result->start);
+            result->stop = _ConvertToBufferCell(result->stop);
+        }
+    }
+    else
+    {
+        // Hyperlink is outside of the current view.
+        // We need to find if there's a pattern at that location.
+        const auto patterns = _activeBuffer().GetPatterns(bufferPos.y, bufferPos.y);
```
Comment (DHowett) [inline src/cascadia/TerminalCore/Terminal.cpp:617]: what happens when a hyperlink starts above the screen but wraps onto another line? Will we accidentally capture only half of it?

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 1 above)
Comment (DHowett) [reply src/cascadia/TerminalCore/Terminal.cpp:629]: I guess I didn't think about how this function was used. It's GOOD that it is robust now, but I suppose now that it was never _called_ with coordinates outside the viewport, eh?

**Review by DHowett — APPROVED** [review]

**Comment by DHowett** [comment]
And hey, does "selectionPointingAtHyperlink" really mean, "selectionPointingAtHyperlinkAfterHavingBeenHavigatedToInMarkMode"? Or, if you manually select a URL with your mouse, will it update to be true???

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 2 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalCore/Terminal.cpp:617]: Welp. Found this out during a bug bash and now it's a line item in #13854. See https://github.com/microsoft/terminal/issues/13854#issuecomment-1233590998 for details on what's going on.

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 1 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalCore/Terminal.cpp:629]: I think given due to https://github.com/microsoft/terminal/issues/13854#issuecomment-1233590998, long-term, we should probably have everything be in the buffer coordinate system. It'll simplify a lot of this extra code. We also need to find hyperlinks outside of the existing viewport. I assume we didn't do that the first time due to performance concerns. :/

**Comment by carlos-zamora** [comment]
> And hey, does "selectionPointingAtHyperlink" really mean, "selectionPointingAtHyperlinkAfterHavingBeenHavigatedToInMarkMode"?

yeah, it really means "selectionPointingAtHyperlinkAfterHavingBeenHavigatedToInMarkMode". In fact, that's really the only scenario we need to keep track of so that we skip over the hyperlink we're currently on.

> Or, if you manually select a URL with your mouse, will it update to be true???

This creates a weird corner case that may require further thinking. Right now, ctrl+enter is a mark mode thing. So if you create a selection using your mouse (and even if you modify it using shift+left/right), ctrl+enter won't do anything. You have to enter mark mode to be able to do try and open it as a URL.

Perhaps the correct solution here is...
```
if (IsSelectionActive() && /*ctrl+enter pressed*/)
{
	if (IsInMarkMode())
	{
		TryOpenURL();
	}
	else if (IsBoundToAction(/*ctrl+enter*/))
	{
		DoAction(/*ctrl+enter*/);
	}
	else
	{
		TryOpenURL();
	}
}
```

Let me know if you agree and I'll move this into #13854

**Pull request merged**